Vehicle register page to db

DoctorDib opened the pull request:
Changes:
 - Added a link between vehicle upload page and the database
 - Added vehicle categories to the upload page
 - Added new SQL function as well as helper function to easily upload vehicle to DB (will require validation to ensure that there are no duplicates).
 - Removed old code from api.js in vehicles folder

DoctorDib committed:
Linked upload page to DB

Changes:
 - Now uploads to the database
 - Added vehicle categories to the upload page
 - Removed old code from api.js in vehicles folder
 - Added new SQL function as well as helper function to easily upload vehicle to DB (will require validation to ensure that there are no duplicates).

DoctorDib committed:
Fixing Lint errors

Changes:
 - Fixed lint errors
 - NOTE: Loads of warnings because of the use of (any) - I am unsure what to replace them with

BenRGH commented on client/src/Common/Enums/VehicleCatehoryEnum.ts:
```diff
@@ -0,0 +1,9 @@
+enum VehicleCategories {
```
you could just put the export default here instead since the file has the name of the enum

DoctorDib replied:
Already tried that, I get the following lint error:
![image](https://user-images.githubusercontent.com/14931616/95660349-20f8cc80-0b1f-11eb-9fb8-6721bd911c96.png)

BenRGH replied:
Fair enough

BenRGH commented on client/src/Common/Interfaces/interfaces.tsx:
```diff
@@ -58,9 +60,10 @@ export interface IInputFields {
-    v5cVerificationYear: string,
-    [key: string]: string | Array<string> | Array<IChip> | Date;
+    v5cVerificationDate: string,
+    [key: string]: string | Array<string> | Array<IChip> | Date | number;
```
Is this just being used as an "everything else"?

DoctorDib replied:
Nope, it's a way to use it in a loop, e.g. object[key]

BenRGH replied:
As in loop over the object?

DoctorDib replied:
Yeah, getting rid of that line will provide link errors

BenRGH commented on client/src/Components/VehicleUploadInputs/index.tsx:
```diff
@@ -43,6 +46,7 @@ const VehicleUploadInputs: React.FC<IImageUploaderProps> = () => {
     const classes: IClasses = styles();
 
     const [input, setInput] = useState<IInputFields>(BlankInputs);
+    const [dateStolen, setDateStolen] = useState<Date>(new Date());
```
Should we consider 'undefined' as an option here since the vehicle might not be stolen yet

DoctorDib replied:
Currently we're only working on the scenario that the vehicle is indeed stolen, perhaps later on we have a toggle that says "Reporting stolen vehicle / Uploading Vehicle"

BenRGH replied:
Good good

BenRGH commented on client/src/Components/VehicleUploadInputs/index.tsx:
```diff
@@ -163,6 +179,28 @@ const VehicleUploadInputs: React.FC<IImageUploaderProps> = () => {
+        AddNewVehicle(1, input);
+    };
+
+    const SetCategoryOptions = Object.keys(VehicleCategoryEnum).map(
```
Missing type

BenRGH replied:
Also this could just be a for each

DoctorDib replied:
That loop could have many different ways, .map works fine. Also the lint didn't pick up the missing type, so I assumed it didn't need it

BenRGH replied:
It doesn't always care about types but you should put them when you can 👍

BenRGH commented on client/src/Components/VehicleUploadInputs/index.tsx:
```diff
@@ -163,6 +179,28 @@ const VehicleUploadInputs: React.FC<IImageUploaderProps> = () => {
+        },
+    );
+
     const SetInputs = Object.keys(input).map((key:string):any => {
```
Another for each, you don't need to use strings for the enums, just use the enums as the cases

DoctorDib replied:
Map works just as well as a the forEach loop, unless there are some advantages that I seem to be missing

BenRGH replied:
It's fine I just don't like the use of Object.keys for everything

BenRGH commented on client/src/Components/VehicleUploadInputs/index.tsx:
```diff
@@ -187,18 +225,36 @@ const VehicleUploadInputs: React.FC<IImageUploaderProps> = () => {
                     </Grid>
                 );
+            case 'category':
+                return (
```
Like I said before, each of these return cases could just be a component or a const by itself, makes them reuseable and discrete

DoctorDib replied:
If you want to do that, then go ahead. I've already attempted it and failed to make it smaller than what current size.

BenRGH replied:
It's not to make it smaller, it's to make it clearer what's going on. A big switch case is hard to deal with especially if we have more fields later on, having individual fields being separate makes it easier to change and understand. At least IMO. If you aren't gonna do it make a task and I will later.

DoctorDib replied:
https://trello.com/c/5GIvZKCE/55-making-the-switch-cases-easier-to-read-understand

BenRGH commented on client/src/Helpers/Blanks.tsx:
```diff
@@ -64,12 +64,13 @@ export const BlankInputs: IInputFields = {
     vin: '',
     make: '',
     model: '',
+    category: 0,
```
I think all these blanks should be renamed to defaultX

BenRGH replied:
Lowercase too since they're not components

DoctorDib replied:
Agreed

DoctorDib replied:
Renamed all "Blanks" to "Defaults" and used lowercasing as suggested

BenRGH commented on client/src/Helpers/DB_Helpers.tsx:
```diff
@@ -86,3 +89,45 @@ export const UpdateVehicleStat = async (vehicleID: number, newStat:number): Prom
+
+    // Extracting the string values from an array of objects
+    // e.g. [{key: 1, value: "one"}, {key: 2, value: "two"}]
+    let featuresArray = SortFeaturesArray(data.featuresArray);
```
This could just be data.featuresArray.map((x) => x.value), no need for a helper func

DoctorDib replied:
Looks messy

BenRGH replied:
It's only a single map method, we don't need a helper just to map especially when the helper has any types

DoctorDib replied:
The helper is there more for possible later uses, which is why it is a helper function.

BenRGH replied:
It's a very specific helper that looks for a .value property

DoctorDib replied:
It's a specific helper function for the features array

BenRGH replied:
Well at least rename it, it doesn't sort anything atm

DoctorDib replied:
Renamed to "ExtractValues"

BenRGH commented on client/src/Helpers/helper.tsx:
```diff
@@ -6,3 +6,19 @@ export const GetDateTime = ():string => {
 
 export const SQLStringProtection = (message:string):string => message.replace(/[\t\r\n]|(--[^\r\n]*)|(\/\*[\w\W]*?(?=\*)\*\/)/gi, '');
+
+export const CheckSQLInObject = (object:any) => {
```
any = bad

DoctorDib replied:
Was trying to make a helper function that doesn't just take in one instance object, if you have suggestions then please comment the code

BenRGH replied:
If there's a library for this then we don't need anyways

DoctorDib replied:
Since we'll be for sure using a library, I am going to resolve this.

BenRGH commented on client/src/Helpers/helper.tsx:
```diff
@@ -6,3 +6,19 @@ export const GetDateTime = ():string => {
 export const SQLStringProtection = (message:string):string => message.replace(/[\t\r\n]|(--[^\r\n]*)|(\/\*[\w\W]*?(?=\*)\*\/)/gi, '');
+
+export const CheckSQLInObject = (object:any) => {
+    const cleanObject = object;
```
newObject makes more sense, but not necessary

DoctorDib replied:
I could argue newObject is an empty object

BenRGH replied:
Context says it isn't

DoctorDib replied:
Changes have been made to "newObject"

BenRGH commented on client/src/Helpers/helper.tsx:
```diff
@@ -6,3 +6,19 @@ export const GetDateTime = ():string => {
 };
 
 export const SQLStringProtection = (message:string):string => message.replace(/[\t\r\n]|(--[^\r\n]*)|(\/\*[\w\W]*?(?=\*)\*\/)/gi, '');
```
Are there not injection libraries that do this for us?

DoctorDib replied:
Good idea, would have more protection as well

BenRGH replied:
And means we don't have to worry, can just dump all the effort on them 😁

DoctorDib replied:
@MidgetJake - would it be worth having these checks to be ran in the Lambda functions (backend) instead of frontend?

DoctorDib replied:
A discussion has been launched on a Trello card that can be found at https://trello.com/c/xXw118Vg/58-install-a-sanitizing-package-for-either-front-or-backend

DoctorDib commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Making the Switch Cases easier to read / understand](https://trello.com/c/5GIvZKCE/55-making-the-switch-cases-easier-to-read-understand)

DoctorDib committed:
Merge branch 'master' into vehicle-register-page-to-db

DoctorDib committed:
PR comment changed implemented

Changes:
 - Renamed all files related to "Blanks" over to "Default" as suggested in a PR comment.
 - Renamed "cleanObject" to "newObject" as requested from PR Comment

DoctorDib commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Attach vehicle upload page to DB](https://trello.com/c/EFEmiKap/53-attach-vehicle-upload-page-to-db)

BenRGH commented on client/src/Helpers/DB_Helpers.tsx:
```diff
@@ -4,7 +4,7 @@ import {
     SQLStringProtection,
     CheckSQLInObject,
-    SortFeaturesArray,
+    SortFeaturesArray as ExtractValue,
```
I feel like this is cheating/malicious compliance but it'll do I guess

BenRGH reviewed (approved):
Might need jake to glaze over the sql stuff but I'm happy for this to go in

DoctorDib commented:
To be honest, the SQL stuff doesn't really have a wrong answer.
